feat: add support for yarn 2 lock files #56

regevbr opened the pull request:
- [x] Tests written and linted
- [x] Documentation written / README.md updated [https://snyk.io/docs/snyk-for-node/](i)
- [x] Follows [CONTRIBUTING agreement](CONTRIBUTING.md)
- [x] Commit history is tidy [https://git-scm.com/book/en/v2/Git-Branching-Rebasing](i)
- [ ] Reviewed by Snyk team

### What this does

Fixes #56
Allows parsing of yarn 2 lock files, which are different from yarn 1

### Notes for the reviewer

- I used `@yarnpkg/parsers` to parse the lock file
- I used @yarnpkg/core for parsing the version descriptors that are in the lock files
- yarn 2 is only supported from node 10 and above (I didn't compromise the compatibility of this library)
- I reused the yarn 1 parser logic, by normalizing the dep tree from yarn 2
- I fixed some existing tests that were wrong
- To realize that the project is using yarn 2, I check the existence of [.yarnrc.yml](https://yarnpkg.com/advanced/migration#update-your-configuration-to-the-new-settings)
- I had to upgrade to typescript 3.4 (the lowest possible version) as the typings in the dependencies of yarns packages were causing build issues and the only way to ignore them (as they are not important) is to use tsconfig `skipLibCheck`, which was only introduced (or at least worked) in typescript 3.4

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/snyk/nodejs-lockfile-parser?pullRequest=57) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/snyk/nodejs-lockfile-parser?pullRequest=57) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/snyk/nodejs-lockfile-parser?pullRequest=57) it.</sub>

regevbr commented:
@orsagie @lili2311 @FauxFaux @miiila can you please CR this?

lili2311 commented:
Hi @regevbr 👋 Thank you was this awesome contribution, we have many team members on holiday today so we would like to discuss this internally with them when they are back tomorrow and get back to you.

regevbr commented:
@lili2311 how are you?

Can you please let me know when you are expected to fix the yarn2 issue? this is the only reason I can't migrate my projects to use yarn2...

regevbr commented:
@arcanis I know you are a very busy man, but I was wondering if you can take a quick look into my PR here to see if I have done it right and if maybe there is a better way to achieve it?
I thought about resuing the code from each relevant plugin but they are too coupled to the actual resolving logic.

arcanis commented on lib/index.ts:
```diff
@@ -79,7 +90,11 @@ async function buildDepTreeFromFiles(
     lockFileType = LockfileType.npm;
   } else if (_.endsWith(lockFilePath, 'yarn.lock')) {
-    lockFileType = LockfileType.yarn;
+    if (fs.existsSync(path.resolve(root, lockFilePath.replace('yarn.lock', '.yarnrc.yml')))) {
+      lockFileType = LockfileType.yarn2;
+    } else {
+      lockFileType = LockfileType.yarn;
+    }
```
Of note, one other way to detect for the 2.x lockfiles is to check whether it contains the `__metadata` string.

regevbr replied:
Thanks I will replace the check

regevbr replied:
Actually, the check here happens before the file is actually parsed, and I can't assume it is a yaml file here, so I think it is best to leave it as is.

arcanis commented on lib/parsers/yarn-lock-parse-base.ts:
```diff
@@ -0,0 +1,165 @@
+        strict = true ): Promise<PkgTree> {
+        if (lockfile.type !== this.type) {
+            throw new InvalidUserInputError('Unsupported lockfile provided. ' +
+                'Please provide `package-lock.json`.');
```
```suggestion
                'Please provide a `yarn.lock` file.');
```

regevbr replied:
I will leave that decision to the maintainers

arcanis commented on lib/parsers/yarn-utils.ts:
```diff
@@ -0,0 +1,72 @@
+                descriptors.push(`${name}@${range.selector}`);
+                descriptors.push(`${name}@${protocol}${range.selector}`);
+                break;
+            case 'http:':
+            case 'https:':
+            case 'link:':
+            case 'portal:':
+            case 'exec:':
+            case 'workspace:':
+            case 'virtual:':
+                descriptors.push(`${name}@${protocol}${range.selector}`);
+                break;
```
An alternative would be to put this in the `default` block - probably more future proof, since users can define their own protocols through the plugin API.

regevbr replied:
Yes it thought about user defined plugins, so you suggest to not throw in case of unknown protocol at all?

arcanis commented on lib/parsers/yarn2-lock-parse.ts:
```diff
@@ -0,0 +1,56 @@
+      throw new UnsupportedRuntimeError('yarn.lock parsing is supported for ' +
+        'Node.js v10 and higher.');
+    }
+    this.yarnLockfileParser = require('@yarnpkg/parsers');
+    const structUtils = require('@yarnpkg/core').structUtils;
```
Why not `import` (if it's intended, like lazy loading or similar, better add a comment to explain why)?

arcanis replied:
Also note that the `@yarnpkg/parsers` is really just a super light wrapper around `js-yaml`. Might make sense to use it directly, your call.

regevbr replied:
Thought about using straight yaml parser but wasn't sure it is enough. I will replace it. 
About the lazy loading it is mentioned in a few places. This library still needs to support node 8 and above

regevbr commented:
@arcanis thanks for the review! much appreciated :-) I will fix all your comments

lili2311 commented on lib/index.ts:
```diff
@@ -79,7 +90,11 @@ async function buildDepTreeFromFiles(
     lockFileType = LockfileType.npm;
   } else if (_.endsWith(lockFilePath, 'yarn.lock')) {
-    lockFileType = LockfileType.yarn;
+    if (fs.existsSync(path.resolve(root, lockFilePath.replace('yarn.lock', '.yarnrc.yml')))) {
```
is there anywhere else these can live? they are always in the same directory?

regevbr replied:
No, they are in the same directory

lili2311 commented on lib/index.ts:
```diff
@@ -79,7 +90,11 @@ async function buildDepTreeFromFiles(
+    if (fs.existsSync(path.resolve(root, lockFilePath.replace('yarn.lock', '.yarnrc.yml')))) {
+      lockFileType = LockfileType.yarn2;
+    } else {
+      lockFileType = LockfileType.yarn;
```
Note: we need a way to send some meta back to track yarn v2 tests vs others

regevbr replied:
The lock file is sent back in the return value of `buildDepTree`. You mean you need to handle the different type in the code that uses that library?

lili2311 commented on lib/index.ts:
```diff
@@ -53,6 +54,16 @@ async function buildDepTree(
+      } else {
+        throw new UnsupportedRuntimeError('Parsing `yarn.lock` is not ' +
+          'supported on Node.js version less than 10. Please upgrade your ' +
+          'Node.js environment or use `package-lock.json`');
```
in this situation an older  v1 `yarn.lock` would also work? Doesn't make sense to ask to switch to `npm`

regevbr replied:
It will not, as the lock file in v1 is not standard yaml file... And in any case yarn2 only works with node 10 and above so I don't think we will have a problem here

dkontorovskyy commented:
@regevbr that is really nice contribution. I will pull it to test locally first and then let's try to merge it. If I will need any of your help, are you up to pairing in close future?

regevbr commented:
@dkontorovskyy sure thing! anything you need. Thanks!

regevbr commented:
I rebased as there were some changes made to master.
I also synced the tree size limits logic to the base yarn parser and added the relevant test to yarn2 parser test

dkontorovskyy commented:
Hey @regevbr. I run local tests on your PR and it looks great 🎉 . I would like to make a graceful roll out for yarn v2, so it would be great if we can pair and separate this PR in 2 smaller bits and take it 1 by 1 into this plugin. I would like to schedule a pairing session with you whenever you free. What do you think about that?

regevbr commented:
@dkontorovskyy thanks!
Sure we can do a session, I'm generally available 12am-3am GMT+3
How do you want to split it?

dkontorovskyy commented:
@regevbr luckily we are in the same timezone 😃 
I can do 12AM-2PM tomorrow. Want me to send you an invite to zoom meeting? How can I contact you ?
I would like to split it into two: first one is refactor needed for a new feature and second one is actually yarn2 support that will be easily added once the first one is in.

dkontorovskyy commented:
@regevbr just keeping update public 😃 . We are looking to plan this feature properly this sprint and start active development next sprint. This is an amazing contribution, so we would try to bring it into the core of Snyk as soon as possible, we just need to make sure this can be done harmlessly to the overall system 🙏

dkontorovskyy commented:
@regevbr first refactor PR went in 💪

regevbr commented:
@dkontorovskyy that's good news!
I will rebase this branch so we will be ready for the next phasse

dkontorovskyy commented:
@arcanis using `@yarnpkg/core` in module breaks `tsc` with error 
```
node_modules/@types/emscripten/index.d.ts:53:33 - error TS2304: Cannot find name 'WebGLRenderingContext'.
53     preinitializedWebGLContext: WebGLRenderingContext;
                                   ~~~~~~~~~~~~~~~~~~~~~
node_modules/@types/emscripten/index.d.ts:67:28 - error TS2304: Cannot find name 'MessageEvent'.
67     onCustomMessage(event: MessageEvent): void;
```

I see that there was an addition to package.json of `@yarnpgk/libzip` to use `@types/emscripten` as production dependencies, because of the bug (https://github.com/yarnpkg/berry/issues/746).

Loos like node.js doesn't like WebAssembly types.

Are you aware of any workarounds for that except silencing TSC compiler with `--skipLibCheck` ?

regevbr commented:
@DaniellaNiceSnyk this is solved by adding "dom" as a lib to tsconfig:
```json
        "lib": [
            "es2017",
            "dom"
        ],
```

I tried getting the library to compile without skipping the lib check, and it requires a lot of changes as there are multiple errors originating from different modules.
The only errors left comes from d.ts reference issues in @yarnpkg/core. For example:
```sh
node_modules/@yarnpkg/core/lib/structUtils.d.ts:91:66 - error TS2307: Cannot find module '../../yarnpkg-fslib/sources'.

91 export declare function slugifyLocator(locator: Locator): import("../../yarnpkg-fslib/sources").Filename;
```

I will open a ticket to fix it in yarn. You can check out all the changes I made in the next commit to this branch and let me know if that is that important for you to not skip the lib check.

regevbr commented:
@dkontorovskyy  I have worked with @arcanis and the issue is now fixed!
There is a new issue in [clipanion](https://github.com/arcanis/clipanion/issues/25) which I will try to get resolved soon as well

regevbr commented:
@dkontorovskyy I fixed the issue in clipanion and now the build works perfectly!
How do we proceed?

dkontorovskyy commented:
@regevbr that is so cool!! What was the issue and fix?

regevbr commented:
@dkontorovskyy in clipanion? see https://github.com/arcanis/clipanion/issues/25 for more details, in high level, the tool used for minifying the d.ts file has bugs in it causing a faulty d.ts file.

If you are asking about the issue in @yarnpkg/core than it is very complicated you can check out my 
 summary at https://github.com/yarnpkg/berry/issues/1278#issuecomment-640903243

dkontorovskyy commented:
@regevbr I merged your PR in our separate feature branch to run some final round of investigation we are doing with the team today. Will keep you posted 🤗

regevbr commented:
Thanks @dkontorovskyy!
Looking forward to hear back from you

regevbr commented:
@dkontorovskyy! how's it going? I see that the feature was finally merged! does that mean I can enable snyk in my yarn2 packages now (ci and PR guard)?

regevbr commented:
@dkontorovskyy! how's it going? I see that the feature was finally merged! does that mean I can enable snyk in my yarn2 packages now (ci and PR guard)?

dkontorovskyy commented:
Hi @regevbr, sorry for spacing out!

It should work in CLI now. Just try scanning your project locally or in CI. Not in PR guard yet. My team will be looking soon into this.

regevbr commented:
That's great news! Please let me know when the guard is working as well as we use it as well :-)

kachkaev commented:
Still getting parse errors in a Github PR here: https://github.com/kachkaev/njt/pull/29

<img width="942" alt="Screenshot 2020-12-24 at 11 35 29" src="https://user-images.githubusercontent.com/608862/103086387-6d12c380-45dc-11eb-9694-ef9dc7a21b26.png">

What are the current plans to resolve this?

I had to replace Snyk integration with [running `yarn audit` on schedule](https://github.com/kachkaev/suppress-exit-code/blob/4568e653092b733ccb912042c03eedf41736c7e1/.github/workflows/dependency-checks.yml) in another project because it was blocking the PR merge. Looking forward to bring Snyk back when CI integration is finally working!

cc @arcanis

kachkaev commented:
👋 from August 2021 😅

https://github.com/kachkaev/njt/pull/29

<img width="923" alt="Screenshot 2021-08-07 at 12 45 52" src="https://user-images.githubusercontent.com/608862/128599037-43821524-4762-41dd-bf8d-21835011c3dd.png">

<img width="1008" alt="Screenshot 2021-08-07 at 12 46 11" src="https://user-images.githubusercontent.com/608862/128599051-0e4437ee-ef22-4acb-bb71-6bb6d0c2a89b.png">

kachkaev commented:
UPD: 👋 from November 2021 😅

https://github.com/snyk/snyk/issues/1518#issuecomment-974822508

<img width="786" alt="Screenshot 2021-11-21 at 13 42 53" src="https://user-images.githubusercontent.com/608862/142764840-e0a92895-25ae-431e-bf15-3b6238aacbaf.png">
